### ISSUE TITLE
Check compatibility with latest version of GEANT4 and possibly consolidate/update make system.

### ISSUE THREAD

**Issue by cwwalter** [WCSim/WCSim#9]
Check to make sure this system builds and compiles with GEANT4 version 10.
Should the make system should be consolidated?  Do we have to use cmake now?

**Comment by aihimmel** [comment]
This could also help with getting WCSim running on the computers in Kamioka.  There are two version of Geant4 there already installed: 4.9.3 and 4.9.6.  The former is too old (no Mie scattering) and the latter is too new (no SetModel in the Optical Boundary code).  On the plus side, it looks like the CMakeLists.txt file already in place will at least allow WCSim to start building.  

In practice, its the deprecated code that's probably going to require the most sorting out.  If I remember correctly, this optical boundary issue is not so much a change in the physics but a change in how you specify which reflection model you want to use.

**Comment by tfeusels** [comment]
Issue #8 relates to the version of Geant4 used, as it would be desirable to document and use up to date PhysicsLists. If @matthewmalek agrees I would start making some branches which compile with Geant 4.9.6 and Geant 4.10.00.p04. So far it seems quite easy to use cmake, together with the Geant4 .cmake files. Just need to update the code, validate the code I think.

**Comment by aihimmeltest** [comment]
It would be of great value to get a version that works with 4.9.6 with minimal changes.  This version of Geant4 is available on the Kamioka computer systems, which would allow a number of new people to get underway.

**Comment by aihimmel** [comment]
It would be of great value to get a version that works with 4.9.6 with minimal changes. This version of Geant4 is available on the Kamioka computer systems, which would allow a number of new people to get underway.

**Comment by tfeusels** [comment]
This would be a one line change to get it running (just tested). Let me try to push my feature branch with this fix once I fix my own https push issues. Validating the changes in physics would take more time (and is the core check of #8 ).

**Comment by cwwalter** [comment]
Did you have to make any physics changes for this?  I think for 4.9.6 normal make should still be fine right? But, I remember something about surfaces changing.   That stuff can be tricky.

**Comment by tfeusels** [comment]
To make it compile and run on the default example, no. I also would encourage cmake from this Geant4 onwards. It's much easier and more natural against Geant4's own cmake system. I haven't tested with normal make.
I'll go through the RELEASE_NOTES to see what changed in more detail.
It would be good to have some basic tests we should do for each G4 version, but I assume @matthewmalek could help with that? I just want to get things compiled, running and PhysicsLists verified for now.

**Comment by cwwalter** [comment]
Right... I eventually definitely want to move to cmake.  I was thinking we would do that for 4.10.  But I'm happy to have the cmake stuff working even now.  But, what did you actually have to change?  I assumed you were referring to something in the optical section were you actually referring to the cmake system?

**Comment by aihimmel** [comment]
By the way, the solution to the https issues is often just to use ssh instead.

**Comment by tfeusels** [comment]
Yes, just uncommenting the SetModel for the Optical section, which is a deprecated option since Geant4.9.6. I now have a precompiler #define set by CMake depending on the Geant4 version that does the trick.
I'll push the code tomorrow.

**Comment by cwwalter** [comment]
Can you seperate any pull requests for cmake stuff and optical stuff.  We want every feature to be handled in its own request.

**Comment by aihimmel** [comment]
Hmmm...I'm not sure having a #define which changes behavior based on which version of Geant we're using is such a good idea.  I'm trying to remember what happened with this SetModel change.  I don't remember if it's a real change in the physics, or just an interface change.  It would be good to see if the behavior is the same before/after removing SetModel within the same Geant version, if possible.

**Comment by cwwalter** [comment]
Ah... I assumed you were referring to the make system (and even then I'm not a fan of the preprocessor but was going to see what you did assuming it would eventually go away).  But, I agree with Alex if you are talking about optical models.

**Comment by tfeusels** [comment]
From the Geant4.9.6 Release Notes:
"Obsolete methods Get/SetModel() have been removed from optical processes, requiring an update (i.e. removal of such calls) for those custom physics lists and physics configurations in which optical boundary process is defined." which is what I did.
If we want WCSim to work with that Geant4 version and earlier versions, the one line #define for the SetModel based on an automated recognition of which Geant4 version is used, is necessary.
If we want to break backwards compatibility, then this is not needed. Currently I would prefer to keep backwards compatibility. Pushing it later today so you can see what I did.

I expect there will be actual physics changes when changing to this or later Geant4 versions, not related to the SetModel, eg. "Hadronic physics now require new G4SAIDDATA data files." and "EM physics (both standard and low-energy) now require the G4EMLOW data files". Should we move that part of the discussion to issue #8 maybe.

**Comment by aihimmel** [comment]
We could probably test the effect using the laser mode.  That way we're starting with optical photons and any other physics changes shouldn't matter, and we can see just the effect of this change in the surface model.

**Comment by cwwalter** [comment]
I thought I remember that even in our  current version (4.9.4) it said it was going to be deprecated soon to give people time to switch to the new way. Does it  not work with 4.9.4?

**Comment by tfeusels** [comment]
More details from going through the Geant4 code:
- G4OpBoundaryProcess uses glisur as default model in the constructor in Geant4.9.4. This is something we change to unified in WCSim now. However, when an OpticalSurface is found, it grabs the OpticalModel from the OpticalSurface class. Hence the Get/SetModel calls in G4OpBoundaryProcess are not really used. I think removing the SetModel calls already in the current WCSim for Geant4.9.4 would be a good solution, as Geant4 does not use that anyway.
- We already define our G4OpticalSurfaceModel as unified in WCSimPhysicsList. So essentially, we are now defining it twice, while Geant4.9.4 only uses the internal one from G4OpticalSurfaceModel. Therefore these calls became redundant in Geant4.9.6.

**Comment by aihimmel** [comment]
OK, thanks for digging into this.  So it looks like this call isn't really doing anything at the moment, so I think taking it out is probably the right solution.  Would you be willing to run the test suite, such as it is, before and after doing that in 4.9.4 and see if the results change?

**Comment by tfeusels** [comment]
Just checked how the OpticalSurface is set by default, because we don't seem to set it ourselves in WCSim, and found (in G4OpticalPhysics) that the default SurfaceModel is "unified", which is what we want. We could manually set this, to be distinguish from "glisur" and "LUT" (Lookup Tables) when we define our surfaces in WCSimConstructMaterials.cc This would avoid all confusion (that others need to look this up in the Geant4 code later), unless we start documenting this default model choice.

**Comment by tfeusels** [comment]
Can you remind me what the test suite is?

**Comment by aihimmel** [comment]
Perhaps "test suite" is too dramatic a name.  There's a directory called "verification-test-scripts" which contains a .mac file to run with both the original WCSim and your modifications, and a root script to compare the simulation outputs.  The [README.md](https://github.com/WCSim/WCSim/blob/develop/verification-test-scripts/README.md) file should have detailed instructions.

**Comment by tfeusels** [comment]
Just ran the tests on original WCSim and the modified one, and as expected they are exactly the same (for G4.9.4), only the modified version also compiles with Geant4.9.6. I'll make a pull request.

**Comment by aihimmel** [comment]
Great!  Thanks for running the test.  Please proceed.

**Comment by cwwalter** [comment]
In the pull request please try to detail the changes in G4 and explanation of our change as fully as possible.  It will be a good resource for the future. Thanks!

**Comment by cwwalter** [comment]
Also remember to separate any make and optical surface requests.

**Comment by tfeusels** [comment]
In the end for this request, I didn't have to change the make. It will compile both with cmake and normal make (as I tested both geant4's). 

For geant4.10.0 they are inevitably entangled. IF we want to keep WCSim compatible with both, we NEED ifdef's, that are passed on to the compiler in CMakeLists.txt (and GNUmakefile).

**Comment by tfeusels** [comment]
@matthewmalek : Feel free to pull the branch for geant4.10.0 from my fork. Only need to remove/replace deprecated models in the WCSimPhysicsList (probably will do it with ifdef's to make sure we can still use older Geant4). Essentially this needs to happen (from RELEASE_NOTES): "Replaced LEP with FTFP+BERT in all inelastic physics constructors".
Updating to 10.1 seems much easier.

**Comment by samshort25** [comment]
@tfeusels: do you have an updated Source_At_Start_T2HK.sh script (and anything else that might be relevant beyond the changes that need to be made to WCSim) that you use with GEANT4 4.10 and the corresponding version of CLHEP (2.2.0.4)? I can make the necessary changes (I could be wrong, but it seems like the directory structure is slightly different to the one you get with the standard Hyper-K release) if you don't :)

**Comment by samshort25** [comment]
Actually, ignore that. I think it might just be the "hack" that's different...

**Comment by eosullivan** [comment]
@matthewmalek, does it seem reasonable if I put this under the WCSim 2.0 milestone, meaning that it would be ready for the next release (which would be in January at the latest, though sooner would be better)? This upgrade to Geant4.10 is becoming increasingly important and it would be nice to set a deadline for it.

**Comment by fdilodov** [comment]
Hi, I just wanted to mention that I am working on it - I got it running on my laptop, and now I have forked a version that I will test in the next few days before submitting a pull request. Actually, I think Sam and I will work on testing it. About the milestone I think it seems reasonable based on the current status, I mean the code runs but I really didn't check the results. Incidentally, I think Matthew is very busy with the new job and stuff, perhaps can the assignee be changed to Sam and myself?

**Comment by eosullivan** [comment]
Ok, sure. Can you confirm with @matthewmalek that he is ok with the switch? I'm not sure if he is getting these GitHub notifications.

**Comment by fdilodov** [comment]
Hi Erin,
sorry I don't think I was very clear in my message. The agreement 
already happened several months ago when Sam  started to work on it (but 
then there were several technical problems that she had). So it is 
already agreed. So, I think we just need to change it from Matthew to 
Sam and myself.
Thanks
Francesca

**Comment by eosullivan** [comment]
Ok, I will switch this to Sam (for some reason you can't assign tasks to multiple people on GitHub). Thanks for looking into this!

**Comment by fdilodov** [comment]
Thanks, Erin.

**Comment by tfeusels** [comment]
Related to the second part of this issue: pull request #125 makes sure CMake builds are possible and documentation (in the README) is up to date.
A few notes: - For Geant 4.9.6 to Geant 4.10.01, the cmake version has to be >= 2.8 (6y old but some computers still only have 2.6 as default).
                     - For Geant 4.10.2: one needs cmake >=3.3.0 (6 months old) with gcc 5.2.0 and gcclib 5.2.0. I haven't tried compiling it without cmake to see whether an older gcc would work.

A question related to the first part of this issue: should we aim for the current most recent, stable Geant4.10 as new default, ie. Geant4.10.2? Probably best?

**Comment by samshort25** [comment]
The version of Geant I'm currently using is 4.10.01.p02; it's the version you can checkout using the scripts in hk-hyperk. Are we happy with this? Or do we need to update the setup scripts to get a more recent version of Geant?

**Comment by eosullivan** [comment]
I would like to use 4.10.2, if possible. There looks to be some updated physics, but maybe even more interesting (for me, anyways) is that there looks to be a new sensitve detector method available where you can have multiple sensitive detectors in the same volume in a more elegant way than I have implemented in our code. Plus, updating our recommended Geant and doing the verification is quite a job, so we should aim to have the most recent version so we don't have to do it again anytime soon.
